File: contracts/src/identity-tenant.ts
/**
 * ATS-0001 contract #1 — IdentityTenant.
 *
 * Platform Keycloak identity/tenant deseninin stable-interface reuse'u
 * (ürün boundary AYRI). Fail-closed: bilinmeyen/eksik token için default
 * tenant ÜRETİLMEZ; tenant kapsamı default-deny'dir (ADR-0002).
 */
import type { ActorId, Outcome, TenantId } from "./types.js";

export interface TenantContext {
  readonly tenantId: TenantId;
  readonly actorId: ActorId;
}

export interface IdentityTenant {
  /**
   * Token'dan tenant bağlamını çözer. Geçersiz/eksik token → fail-closed
   * (UNAUTHENTICATED / DENIED); asla default tenant üretmez.
   */
  resolveTenant(token: string): Outcome<TenantContext>;

  /**
   * Bir kaynağın tenant'ı, çağıran bağlamının tenant'ıyla eşleşmiyorsa
   * TENANT_SCOPE_VIOLATION döner (default-deny). Eşleşiyorsa ok(void).
   */
  assertTenantScope(ctx: TenantContext, resourceTenantId: TenantId): Outcome<void>;
}

File: backend/contracts-java/src/main/java/com/ats/contracts/IdentityTenant.java
package com.ats.contracts;

import com.ats.kernel.Ids.ActorId;
import com.ats.kernel.Ids.TenantId;
import com.ats.kernel.Outcome;

/**
 * ATS-0001 #1 IdentityTenant (TS mirror). Fail-closed: bilinmeyen token için
 * default tenant ÜRETİLMEZ; tenant kapsamı default-deny (ATS-0002).
 */
public interface IdentityTenant {

    record TenantContext(TenantId tenantId, ActorId actorId) {}

    /** Geçersiz/eksik token → fail-closed (UNAUTHENTICATED/DENIED); default tenant yok. */
    Outcome<TenantContext> resolveTenant(String token);

    /** Kaynak tenant'ı bağlamla eşleşmezse TENANT_SCOPE_VIOLATION (default-deny). */
    Outcome<Void> assertTenantScope(TenantContext ctx, TenantId resourceTenantId);
}
